### PR TITLE
Features/dns controller

### PR DESCRIPTION
**Pull request by River-sh**
#### What this PR does / why we need it:
The DNS Controller maintains a record of the NodeName mapping to the Raven L7 Proxy IP address in a configmap and mounts it into CoreDNS to support cross pool communication

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/controller/raven/dns/dns_controller.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			ip := net.ParseIP(addr.Address)
+			if ip != nil {
+				allIPs = append(allIPs, ip)
+				break
+			}
+		}
+	}
+	for _, addr := range node.Status.Addresses {
```
Comment (fengshunli) [inline pkg/controller/raven/dns/dns_controller.go:252]: reuse the same for loop, add else if to handle the same logic, reduce redundant code

**Review by fengshunli — CHANGES_REQUESTED** [review]

File: pkg/controller/apis/config/types.go [hunk 2]
```diff
@@ -35,7 +36,7 @@ type YurtManagerConfiguration struct {
 	NodePoolController nodepoolconfig.NodePoolControllerConfiguration
 
 	// GatewayControllerConfiguration holds configuration for GatewayController related features.
-	GatewayController gatewayconfig.GatewayControllerConfiguration
+	GatewayController gatewayconfig.GatewayPickupControllerConfiguration
```
Comment (fengshunli) [inline pkg/controller/apis/config/types.go:39]: renmae GatewayPickupController

File: cmd/yurt-manager/app/options/ravendnscontroller.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		return
+	}
+
+	//fs.BoolVar(&n.CreateDefaultPool, "create-default-pool", n.CreateDefaultPool, "Create default cloud/edge pools if indicated.")
+}
+
+// ApplyTo fills up nodepool config with options.
+func (o *RavenDNSControllerOptions) ApplyTo(cfg *config.RavenDNSControllerConfiguration) error {
+	if o == nil {
+		return nil
```
Comment (fengshunli) [inline cmd/yurt-manager/app/options/ravendnscontroller.go:47]: return err

File: pkg/controller/raven/dns/dns_controller.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+			if ip != nil {
+				allIPs = append(allIPs, ip)
+				break
+			}
+		}
+	}
+	if len(allIPs) == 0 {
+		return "", fmt.Errorf("host IP unknown; known addresses: %v", node.Status.Addresses)
+	}
+	return allIPs[0].String(), nil
```
Comment (fengshunli) [inline pkg/controller/raven/dns/dns_controller.go:264]: add nil value judgment

**Review by fengshunli — CHANGES_REQUESTED** [review]

**Comment by rambohe-ch** [comment]
@River-sh I think it is not a good idea to add several controllers in only one pull request.  because it is not easy to review so much code in only one pull request. would you like to split this pull request and post one controller one by one?

**Pull request closed without merging**